### ISSUE TITLE
OAuth: Support client_secret_jwt for oauth providers when doing token exchange

### ISSUE THREAD

**Issue by kalleep** [grafana/grafana#95168]
**Why is this needed**:
Right now when oauth is setup through any provider we always perform code to token exchange using `client_secret_post`. We will do a post request to the token endpoint and always include `client_secret` in the request url.

We should implement `client_secret_jwt`. What that means is that we use the client secret to sign a jwt that the IPD also can validate using the shared client secret. More details on it can be found here https://openid.net/specs/openid-connect-core-1_0.html#ClientAuthentication.

We would need to look through our providers and see who can support it. 

~It also does not seems like the [oauth2 package we use support it yet](https://github.com/golang/oauth2/issues/744) so either we wait for the pr to be merged or workaround it for now.~

I imagine 1 new config option for providers `client_authentication` that can have two values, either `client_secret_post` (should be default) and `client_secret_jwt`.

On top of these we need to update places in the ui where we can do oauth configuration.

When doing the exchange we should perform the check on what client authentication method is used in https://github.com/grafana/grafana/blob/main/pkg/login/social/connectors/social_base.go#L92-L97 and handle it correctly.

**Comment by naizerjohn-ms** [comment]
Hello Kalle! That is a great explanation of the issue at hand, and I agree with the recommendation to implement two different config options for providers: `client_secret_post` and `client_secret_jwt`.

To provide clarity on if the oauth2 package supports this, they actually do support using `client_secret_jwt`. One of the parameters passed in the Exchange function [here](https://github.com/grafana/grafana/blob/main/pkg/login/social/connectors/social_base.go#L92-L97) is `var opts []oauth2.AuthCodeOption`, where you can add the necessary HTTP parameters for the `client_assertion` flow. Here is a snippet of our current code implementation doing this:

```
opts = append(opts, oauth2.SetAuthURLParam("client_assertion_type", "urn:ietf:params:oauth:client-assertion-type:jwt-bearer"))
opts = append(opts, oauth2.SetAuthURLParam("client_assertion", clientAssertion))
```

What's super neat is that in the oauth2 package, it writes all the opts in this array to the HTTP request, and **only** adds the `client_secret` if it is not null/empty! Here is where this happens in the oauth2 package: [link](https://github.com/golang/oauth2/blob/3e6480915d39dd1a80fa460e56413857f02cc1b9/internal/token.go#L189-L191)

Moving forward, we will work on adding/changing the configs to where there are two values: `client_secret_post` and `client_secret_jwt` in the backend. I will push any commits to this issue so that you and others may provide more guidance on refactoring/verifying the changes are correct. Thanks!

**Comment by kalleep** [comment]
@naizerjohn-ms if you want to get feedback on you work as you go I suggest you open a draft pr, that way its easy for me to track changes as you go

**Comment by naizerjohn-ms** [comment]
@kalleep Should I create a new fork for my development branch? It seems that I cannot publish a new branch directly to grafana/grafana.

**Comment by kalleep** [comment]
Yes you should create a fork

**Comment by naizerjohn-ms** [comment]
I'm going to tackle the backend first since that is what we have working so far. At some point today I will create an open draft pr for my forked repo and link this issue to it

**Comment by naizerjohn-ms** [comment]
@kalleep Please link [this](https://github.com/grafana/grafana/pull/95455) PR to this issue, as this is the draft PR we will work off of. Thanks!

**Comment by naizerjohn-ms** [comment]
@kalleep I can go ahead and create a separate issue pertaining to the token refresh issue unless you want to keep this one open and add it here?

**Comment by mgyongyosi** [comment]
@naizerjohn-ms Yes, please create a separate issue for token refresh and link it to this. Thanks!

**Comment by naizerjohn-ms** [comment]
@mgyongyosi sounds good! Also, I created another PR for Grafana to add support for passing those new flags to the Grafana Azure SDK and it can be found [here](https://github.com/grafana/grafana/pull/98703). All new context needed is in that PR. Not sure if you would be able to expedite this one since you already have a ton of context, but any help is greatly appreciated

**Comment by naizerjohn-ms** [comment]
@mgyongyosi here is the issue i just created to track this bug: [issue](https://github.com/grafana/grafana/issues/98919)
